README.txt
Git OS X Installer
==================

Patches (see the README_PATCHES.txt file) to build a version of Git for
Mac OS X 10.4.8 and later (PPC/X86/32/64) that includes the Git translations
(optionally selected based on System Preferences > International > Languages)
and integrates well with the native OS X libraries.

The built version of Git uses Secure Transport (via a very enhanced version of
curl's darwinssl backend) to fully support SHA-256 https certificates on
Mac OS X 10.4 and Mac OS X 10.5 without need for a custom build of OpenSSL.

All certificates for https/smtps/imaps verification come from the standard
system keychain locations (unless one of the Git certificate configuration
options is used).

Patches are also included to allow GnuPG to generate larger RSA keys (again,
see README_PATCHES.txt for details).

Pre-built installers that include Git, TopGit (optional), the curl comand line
utility (optional) and GnuPG (optional) can be found on the Git OS X Installer
home page.

See the Git OS X Installer home page at:
<http://mackyle.github.io/git-osx-installer>


ChangeLog.txt
=============================
Git OS X Installer Change Log
=============================


------------------------------------------
Git 2.11.1 Installer - released 2017-04-07
------------------------------------------

* Dropped libgcc dylib dependency for even more compatible binaries

* Git v2.11.1
  - update from version 2.10.2
  - real git-p4 now included (was previously inadvertently omitted)
    (a git-p4 compatible version of python must be in $PATH to use it)

* cURL/libcURL 7.53.1
  - update from version 7.51.0

* PCRE 8.40
  - update from version 8.37
  - replace Git's NO_REGEX build option with pcreposix-compat instead
    * Git's NO_REGEX compatibility code is best avoided
    * See <https://github.com/mackyle/pcreposix-compat> for more info

* zlib 1.2.11
  - update from version 1.2.8

* TopGit 0.19.7
  - update from version 0.19.3


------------------------------------------
Git 2.10.2 Installer - released 2016-11-08
------------------------------------------

* Git v2.10.2
  - update from version 2.9.3

* cURL/libcURL 7.51.0
  - update from version 7.50.1
  - drop IDN support
    * as of cURL 7.51.0 IDN support requires libidn2
    * libidn (cURL < 7.51.0) was available as GPL2
    * libidn2 is only available as GPL3 but Git is GPL2

* GnuPG 1.4.21
  - update from version 1.4.19


-----------------------------------------
Git 2.9.3 Installer - released 2016-08-13
-----------------------------------------

* Git v2.9.3
  - update from version 2.8.4

* cURL/libcURL 7.50.1
  - update from version 7.49.1


-----------------------------------------
Git 2.8.4 Installer - released 2016-06-19
-----------------------------------------

* Git v2.8.4
  - update from version 2.8.3

* cURL/libcURL 7.49.1
  - update from version 7.49.0

* libexpat 2.1.1
  - update from version 2.1.0


-----------------------------------------
Git 2.8.3 Installer - released 2016-05-21
-----------------------------------------

* Git v2.8.3
  - update from version 2.8.2

* cURL/libcURL 7.49.0
  - update from version 7.48.0


-----------------------------------------
Git 2.8.2 Installer - released 2016-04-30
-----------------------------------------

* Git v2.8.2
  - update from version 2.7.4
  - drop notes anywhere patch (replacement in 2.8.0)

* cURL/libcURL 7.48.0
  - update from version 7.47.1


-----------------------------------------
Git 2.7.4 Installer - released 2016-03-28
-----------------------------------------

* Git v2.7.4
  - update from version 2.7.2


-----------------------------------------
Git 2.6.6 Installer - released 2016-03-28
-----------------------------------------

* Git v2.6.6
  - update from version 2.6.5

* TopGit 0.19.3
  - update from version 0.19.2


-----------------------------------------
Git 2.5.5 Installer - released 2016-03-27
-----------------------------------------

* Git v2.5.5
  - update from version 2.5.4

* cURL/libcURL 7.47.1
  - update from version 7.46.0

* TopGit 0.19.3
  - update from version 0.19.1


------------------------------------------
Git 2.4.11 Installer - released 2016-03-27
------------------------------------------

* Git v2.4.11
  - update from version 2.4.10
  - git-mw and git-remote-mediawiki now work with https URLs and also on 10.4.x
  - include contrib/git-log-times alternative git log --online utility

* cURL/libcURL 7.47.1
  - update from version 7.45.0
  - eliminate SSL session ID reuse failures on 10.4.x

* lighttpd 1.4.39
  - update from version 1.4.37
  - drop darwin sendfile patch (included in 1.4.38)
  - drop darwin test shutdown patch (included in 1.4.38)
  - drop darwin test crypt md5 patch (included in 1.4.38)
  - drop crypt status patch (included in 1.4.38)
  - drop zero sockaddr structs patch (included in 1.4.38)

* TopGit 0.19.3
  - update from version 0.18


-----------------------------------------
Git 2.7.2 Installer - released 2016-02-29
-----------------------------------------

* Git v2.7.2
  - update from version 2.6.5

* TopGit 0.19.3
  - update from version 0.19.2


-----------------------------------------
Git 2.6.5 Installer - released 2016-02-24
-----------------------------------------

* Git v2.6.5
  - update from version 2.5.4

* cURL/libcURL 7.47.1
  - update from version 7.46.0

* TopGit 0.19.2
  - update from version 0.19.1


-----------------------------------------
Git 2.5.4 Installer - released 2016-02-05
-----------------------------------------

* Git v2.5.4
  - update from version 2.4.10
  - git-mw and git-remote-mediawiki now work with https URLs and also on 10.4.x
  - include contrib/git-log-times alternative git log --online utility

* cURL/libcURL 7.46.0
  - update from version 7.45.0
  - eliminate SSL session ID reuse failures on 10.4.x

* lighttpd 1.4.39
  - update from version 1.4.37
  - drop darwin sendfile patch (included in 1.4.38)
  - drop darwin test shutdown patch (included in 1.4.38)
  - drop darwin test crypt md5 patch (included in 1.4.38)
  - drop crypt status patch (included in 1.4.38)
  - drop zero sockaddr structs patch (included in 1.4.38)

* TopGit 0.19.1
  - update from version 0.18


------------------------------------------
Git 2.4.10 Installer - released 2015-10-27
------------------------------------------

* Git v2.4.10
  - update from version 2.3.10
  - the new ru.po translations work properly in all their plural forms
  - include git-mw and git-remote-mediawiki from contrib/mw-to-git
  - drop thread-utils.c fix (included in v2.4.0)
  - drop git-gui Tcl 8.4 fix (included in v2.4.0)
  - drop contrib/pre-auto-gc-battery fix backport (included in v2.4.0)
  - drop git-instaweb from subdir fix (included in v2.4.0)


------------------------------------------
Git 2.3.10 Installer - reissued 2015-10-16
------------------------------------------

* Installer
  - Replace corrupted .DS_Store file causing  bad folder layout
    (There was nothing wrong with the Git version it installed,
     but since it's being reissued, might as well include updates.)

* Git v2.3.10

src/ccccrypt.c
/* LibTomCrypt, modular cryptographic library -- Tom St Denis
 *
 * LibTomCrypt is a library that provides various cryptographic
 * algorithms in a highly modular and flexible manner.
 *
 * The library is free for all purposes without any express
 * guarantee it works.
 *
 * Tom St Denis, tomstdenis@gmail.com, http://libtom.org
 */

#include <stdint.h>
#include <string.h>

#define XMEMCPY memcpy
#define XMEMCMP memcmp

typedef uint32_t ulong32;

#define LOAD32H(x, y) \
     { x = ((ulong32)((y)[0] & 255)<<24) | \
           ((ulong32)((y)[1] & 255)<<16) | \
           ((ulong32)((y)[2] & 255)<<8)  | \
           ((ulong32)((y)[3] & 255)); }

#define STORE32H(x, y) \
     { (y)[0] = (unsigned char)(((x)>>24)&255); (y)[1] = (unsigned char)(((x)>>16)&255); \
       (y)[2] = (unsigned char)(((x)>>8)&255); (y)[3] = (unsigned char)((x)&255); }

#define ROLc(x, y) ( (((ulong32)(x)<<(ulong32)((y)&31)) | \
                     (((ulong32)(x)&0xFFFFFFFFU)>>(ulong32)(32-((y)&31)))) & 0xFFFFFFFFU )

#define RORc(x, y) ( ((((ulong32)(x)&0xFFFFFFFFU)>>(ulong32)((y)&31)) | \
                     ((ulong32)(x)<<(ulong32)(32-((y)&31)))) & 0xFFFFFFFFU )

#define LTC_ARGCHK(x)
#define CRYPT_OK 0
#define CRYPT_INVALID_ROUNDS -1
#define CRYPT_INVALID_KEYSIZE -1

struct des_key {
    ulong32 ek[32], dk[32];
};

typedef union Symmetric_key {
   struct des_key des;
} symmetric_key;

/**
  @file des.c
  LTC_DES code submitted by Dobes Vandermeer
*/

#define EN0 0
#define DE1 1

static const ulong32 bytebit[8] =
{
    0200, 0100, 040, 020, 010, 04, 02, 01
};

static const ulong32 bigbyte[24] =
{
    0x800000UL,  0x400000UL,  0x200000UL,  0x100000UL,
    0x80000UL,   0x40000UL,   0x20000UL,   0x10000UL,
    0x8000UL,    0x4000UL,    0x2000UL,    0x1000UL,
    0x800UL,     0x400UL,     0x200UL,     0x100UL,
    0x80UL,      0x40UL,      0x20UL,      0x10UL,
    0x8UL,       0x4UL,       0x2UL,       0x1L
};

/* Use the key schedule specific in the standard (ANSI X3.92-1981) */

static const unsigned char pc1[56] = {
    56, 48, 40, 32, 24, 16,  8,  0, 57, 49, 41, 33, 25, 17,
     9,  1, 58, 50, 42, 34, 26, 18, 10,  2, 59, 51, 43, 35,
    62, 54, 46, 38, 30, 22, 14,  6, 61, 53, 45, 37, 29, 21,
    13,  5, 60, 52, 44, 36, 28, 20, 12,  4, 27, 19, 11,  3
};

static const unsigned char totrot[16] = {
    1,   2,  4,  6,
    8,  10, 12, 14,
    15, 17, 19, 21,
    23, 25, 27, 28
};

static const unsigned char pc2[48] = {
    13, 16, 10, 23,  0,  4,      2, 27, 14,  5, 20,  9,
    22, 18, 11,  3, 25,  7,     15,  6, 26, 19, 12,  1,
    40, 51, 30, 36, 46, 54,     29, 39, 50, 44, 32, 47,
    43, 48, 38, 55, 33, 52,     45, 41, 49, 35, 28, 31
};

static const ulong32 SP1[64] =
{
    0x01010400UL, 0x00000000UL, 0x00010000UL, 0x01010404UL,
    0x01010004UL, 0x00010404UL, 0x00000004UL, 0x00010000UL,
    0x00000400UL, 0x01010400UL, 0x01010404UL, 0x00000400UL,
    0x01000404UL, 0x01010004UL, 0x01000000UL, 0x00000004UL,
    0x00000404UL, 0x01000400UL, 0x01000400UL, 0x00010400UL,
    0x00010400UL, 0x01010000UL, 0x01010000UL, 0x01000404UL,
    0x00010004UL, 0x01000004UL, 0x01000004UL, 0x00010004UL,
    0x00000000UL, 0x00000404UL, 0x00010404UL, 0x01000000UL,
    0x00010000UL, 0x01010404UL, 0x00000004UL, 0x01010000UL,
    0x01010400UL, 0x01000000UL, 0x01000000UL, 0x00000400UL,
    0x01010004UL, 0x00010000UL, 0x00010400UL, 0x01000004UL,
    0x00000400UL, 0x00000004UL, 0x01000404UL, 0x00010404UL,
    0x01010404UL, 0x00010004UL, 0x01010000UL, 0x01000404UL,
    0x01000004UL, 0x00000404UL, 0x00010404UL, 0x01010400UL,
    0x00000404UL, 0x01000400UL, 0x01000400UL, 0x00000000UL,
    0x00010004UL, 0x00010400UL, 0x00000000UL, 0x01010004UL
};

static const ulong32 SP2[64] =
{
    0x80108020UL, 0x80008000UL, 0x00008000UL, 0x00108020UL,
    0x00100000UL, 0x00000020UL, 0x80100020UL, 0x80008020UL,
    0x80000020UL, 0x80108020UL, 0x80108000UL, 0x80000000UL,
    0x80008000UL, 0x00100000UL, 0x00000020UL, 0x80100020UL,
    0x00108000UL, 0x00100020UL, 0x80008020UL, 0x00000000UL,
    0x80000000UL, 0x00008000UL, 0x00108020UL, 0x80100000UL,
    0x00100020UL, 0x80000020UL, 0x00000000UL, 0x00108000UL,
    0x00008020UL, 0x80108000UL, 0x80100000UL, 0x00008020UL,
    0x00000000UL, 0x00108020UL, 0x80100020UL, 0x00100000UL,
    0x80008020UL, 0x80100000UL, 0x80108000UL, 0x00008000UL,
    0x80100000UL, 0x80008000UL, 0x00000020UL, 0x80108020UL,
    0x00108020UL, 0x00000020UL, 0x00008000UL, 0x80000000UL,
    0x00008020UL, 0x80108000UL, 0x00100000UL, 0x80000020UL,
    0x00100020UL, 0x80008020UL, 0x80000020UL, 0x00100020UL,
    0x00108000UL, 0x00000000UL, 0x80008000UL, 0x00008020UL,
    0x80000000UL, 0x80100020UL, 0x80108020UL, 0x00108000UL
};

static const ulong32 SP3[64] =
{
    0x00000208UL, 0x08020200UL, 0x00000000UL, 0x08020008UL,
    0x08000200UL, 0x00000000UL, 0x00020208UL, 0x08000200UL,
    0x00020008UL, 0x08000008UL, 0x08000008UL, 0x00020000UL,
    0x08020208UL, 0x00020008UL, 0x08020000UL, 0x00000208UL,
    0x08000000UL, 0x00000008UL, 0x08020200UL, 0x00000200UL,
    0x00020200UL, 0x08020000UL, 0x08020008UL, 0x00020208UL,
    0x08000208UL, 0x00020200UL, 0x00020000UL, 0x08000208UL,
    0x00000008UL, 0x08020208UL, 0x00000200UL, 0x08000000UL,
    0x08020200UL, 0x08000000UL, 0x00020008UL, 0x00000208UL,
    0x00020000UL, 0x08020200UL, 0x08000200UL, 0x00000000UL,
    0x00000200UL, 0x00020008UL, 0x08020208UL, 0x08000200UL,
    0x08000008UL, 0x00000200UL, 0x00000000UL, 0x08020008UL,
    0x08000208UL, 0x00020000UL, 0x08000000UL, 0x08020208UL,
    0x00000008UL, 0x00020208UL, 0x00020200UL, 0x08000008UL,
    0x08020000UL, 0x08000208UL, 0x00000208UL, 0x08020000UL,
    0x00020208UL, 0x00000008UL, 0x08020008UL, 0x00020200UL
};

static const ulong32 SP4[64] =
{
    0x00802001UL, 0x00002081UL, 0x00002081UL, 0x00000080UL,
    0x00802080UL, 0x00800081UL, 0x00800001UL, 0x00002001UL,
    0x00000000UL, 0x00802000UL, 0x00802000UL, 0x00802081UL,
    0x00000081UL, 0x00000000UL, 0x00800080UL, 0x00800001UL,
    0x00000001UL, 0x00002000UL, 0x00800000UL, 0x00802001UL,
    0x00000080UL, 0x00800000UL, 0x00002001UL, 0x00002080UL,
    0x00800081UL, 0x00000001UL, 0x00002080UL, 0x00800080UL,
    0x00002000UL, 0x00802080UL, 0x00802081UL, 0x00000081UL,
    0x00800080UL, 0x00800001UL, 0x00802000UL, 0x00802081UL,
    0x00000081UL, 0x00000000UL, 0x00000000UL, 0x00802000UL,
    0x00002080UL, 0x00800080UL, 0x00800081UL, 0x00000001UL,
    0x00802001UL, 0x00002081UL, 0x00002081UL, 0x00000080UL,
    0x00802081UL, 0x00000081UL, 0x00000001UL, 0x00002000UL,
    0x00800001UL, 0x00002001UL, 0x00802080UL, 0x00800081UL,
    0x00002001UL, 0x00002080UL, 0x00800000UL, 0x00802001UL,
    0x00000080UL, 0x00800000UL, 0x00002000UL, 0x00802080UL
};

static const ulong32 SP5[64] =
{
    0x00000100UL, 0x02080100UL, 0x02080000UL, 0x42000100UL,
    0x00080000UL, 0x00000100UL, 0x40000000UL, 0x02080000UL,
    0x40080100UL, 0x00080000UL, 0x02000100UL, 0x40080100UL,
    0x42000100UL, 0x42080000UL, 0x00080100UL, 0x40000000UL,
    0x02000000UL, 0x40080000UL, 0x40080000UL, 0x00000000UL,
    0x40000100UL, 0x42080100UL, 0x42080100UL, 0x02000100UL,
    0x42080000UL, 0x40000100UL, 0x00000000UL, 0x42000000UL,
    0x02080100UL, 0x02000000UL, 0x42000000UL, 0x00080100UL,
    0x00080000UL, 0x42000100UL, 0x00000100UL, 0x02000000UL,
    0x40000000UL, 0x02080000UL, 0x42000100UL, 0x40080100UL,
    0x02000100UL, 0x40000000UL, 0x42080000UL, 0x02080100UL,
    0x40080100UL, 0x00000100UL, 0x02000000UL, 0x42080000UL,
    0x42080100UL, 0x00080100UL, 0x42000000UL, 0x42080100UL,
    0x02080000UL, 0x00000000UL, 0x40080000UL, 0x42000000UL,
    0x00080100UL, 0x02000100UL, 0x40000100UL, 0x00080000UL,
    0x00000000UL, 0x40080000UL, 0x02080100UL, 0x40000100UL
};

static const ulong32 SP6[64] =
{
    0x20000010UL, 0x20400000UL, 0x00004000UL, 0x20404010UL,
    0x20400000UL, 0x00000010UL, 0x20404010UL, 0x00400000UL,
    0x20004000UL, 0x00404010UL, 0x00400000UL, 0x20000010UL,
    0x00400010UL, 0x20004000UL, 0x20000000UL, 0x00004010UL,
    0x00000000UL, 0x00400010UL, 0x20004010UL, 0x00004000UL,
    0x00404000UL, 0x20004010UL, 0x00000010UL, 0x20400010UL,
    0x20400010UL, 0x00000000UL, 0x00404010UL, 0x20404000UL,
    0x00004010UL, 0x00404000UL, 0x20404000UL, 0x20000000UL,
    0x20004000UL, 0x00000010UL, 0x20400010UL, 0x00404000UL,
    0x20404010UL, 0x00400000UL, 0x00004010UL, 0x20000010UL,
    0x00400000UL, 0x20004000UL, 0x20000000UL, 0x00004010UL,
    0x20000010UL, 0x20404010UL, 0x00404000UL, 0x20400000UL,
    0x00404010UL, 0x20404000UL, 0x00000000UL, 0x20400010UL,
    0x00000010UL, 0x00004000UL, 0x20400000UL, 0x00404010UL,
    0x00004000UL, 0x00400010UL, 0x20004010UL, 0x00000000UL,
    0x20404000UL, 0x20000000UL, 0x00400010UL, 0x20004010UL
};

static const ulong32 SP7[64] =
{
    0x00200000UL, 0x04200002UL, 0x04000802UL, 0x00000000UL,
    0x00000800UL, 0x04000802UL, 0x00200802UL, 0x04200800UL,
    0x04200802UL, 0x00200000UL, 0x00000000UL, 0x04000002UL,
    0x00000002UL, 0x04000000UL, 0x04200002UL, 0x00000802UL,
    0x04000800UL, 0x00200802UL, 0x00200002UL, 0x04000800UL,
    0x04000002UL, 0x04200000UL, 0x04200800UL, 0x00200002UL,
    0x04200000UL, 0x00000800UL, 0x00000802UL, 0x04200802UL,
    0x00200800UL, 0x00000002UL, 0x04000000UL, 0x00200800UL,
    0x04000000UL, 0x00200800UL, 0x00200000UL, 0x04000802UL,
    0x04000802UL, 0x04200002UL, 0x04200002UL, 0x00000002UL,
    0x00200002UL, 0x04000000UL, 0x04000800UL, 0x00200000UL,
    0x04200800UL, 0x00000802UL, 0x00200802UL, 0x04200800UL,
    0x00000802UL, 0x04000002UL, 0x04200802UL, 0x04200000UL,
    0x00200800UL, 0x00000000UL, 0x00000002UL, 0x04200802UL,
    0x00000000UL, 0x00200802UL, 0x04200000UL, 0x00000800UL,
    0x04000002UL, 0x04000800UL, 0x00000800UL, 0x00200002UL
};

static const ulong32 SP8[64] =
{
    0x10001040UL, 0x00001000UL, 0x00040000UL, 0x10041040UL,
    0x10000000UL, 0x10001040UL, 0x00000040UL, 0x10000000UL,
    0x00040040UL, 0x10040000UL, 0x10041040UL, 0x00041000UL,
    0x10041000UL, 0x00041040UL, 0x00001000UL, 0x00000040UL,
    0x10040000UL, 0x10000040UL, 0x10001000UL, 0x00001040UL,
    0x00041000UL, 0x00040040UL, 0x10040040UL, 0x10041000UL,
    0x00001040UL, 0x00000000UL, 0x00000000UL, 0x10040040UL,
    0x10000040UL, 0x10001000UL, 0x00041040UL, 0x00040000UL,
    0x00041040UL, 0x00040000UL, 0x10041000UL, 0x00001000UL,
    0x00000040UL, 0x10040040UL, 0x00001000UL, 0x00041040UL,
    0x10001000UL, 0x00000040UL, 0x10000040UL, 0x10040000UL,
    0x10040040UL, 0x10000000UL, 0x00040000UL, 0x10001040UL,
    0x00000000UL, 0x10041040UL, 0x00040040UL, 0x10000040UL,
    0x10040000UL, 0x10001000UL, 0x10001040UL, 0x00000000UL,
    0x10041040UL, 0x00041000UL, 0x00041000UL, 0x00001040UL,
    0x00001040UL, 0x00040040UL, 0x10000000UL, 0x10041000UL
};

static void cookey(const ulong32 *raw1, ulong32 *keyout);

static void deskey(const unsigned char *key, short edf, ulong32 *keyout)
{
    ulong32 i, j, l, m, n, kn[32];
    unsigned char pc1m[56], pcr[56];

    for (j=0; j < 56; j++) {
        l = (ulong32)pc1[j];
        m = l & 7;
        pc1m[j] = (unsigned char)((key[l >> 3U] & bytebit[m]) == bytebit[m] ? 1 : 0);
    }

    for (i=0; i < 16; i++) {
        if (edf == DE1) {
           m = (15 - i) << 1;
        } else {
           m = i << 1;
        }
        n = m + 1;
        kn[m] = kn[n] = 0L;
        for (j=0; j < 28; j++) {
            l = j + (ulong32)totrot[i];
            if (l < 28) {
               pcr[j] = pc1m[l];
            } else {
               pcr[j] = pc1m[l - 28];
            }
        }
        for (/*j = 28*/; j < 56; j++) {
            l = j + (ulong32)totrot[i];
            if (l < 56) {
               pcr[j] = pc1m[l];
            } else {
               pcr[j] = pc1m[l - 28];
            }
        }
        for (j=0; j < 24; j++)  {
            if ((int)pcr[(int)pc2[j]] != 0) {
               kn[m] |= bigbyte[j];
            }
            if ((int)pcr[(int)pc2[j+24]] != 0) {
               kn[n] |= bigbyte[j];
            }
        }
    }

    cookey(kn, keyout);
}

static void cookey(const ulong32 *raw1, ulong32 *keyout)
{
    ulong32 *cook;
    const ulong32 *raw0;
    ulong32 dough[32];
    int i;

    cook = dough;
    for(i=0; i < 16; i++, raw1++)
    {
        raw0 = raw1++;
        *cook    = (*raw0 & 0x00fc0000L) << 6;
        *cook   |= (*raw0 & 0x00000fc0L) << 10;
        *cook   |= (*raw1 & 0x00fc0000L) >> 10;
        *cook++ |= (*raw1 & 0x00000fc0L) >> 6;
        *cook    = (*raw0 & 0x0003f000L) << 12;
        *cook   |= (*raw0 & 0x0000003fL) << 16;
        *cook   |= (*raw1 & 0x0003f000L) >> 4;
        *cook++ |= (*raw1 & 0x0000003fL);
    }

    XMEMCPY(keyout, dough, sizeof dough);
}

static void desfunc(ulong32 *block, const ulong32 *keys)
{
    ulong32 work, right, leftt;
    int cur_round;

    leftt = block[0];
    right = block[1];

    work = ((leftt >> 4)  ^ right) & 0x0f0f0f0fL;
    right ^= work;
    leftt ^= (work << 4);

    work = ((leftt >> 16) ^ right) & 0x0000ffffL;
    right ^= work;
    leftt ^= (work << 16);

    work = ((right >> 2)  ^ leftt) & 0x33333333L;
    leftt ^= work;
    right ^= (work << 2);

    work = ((right >> 8)  ^ leftt) & 0x00ff00ffL;
    leftt ^= work;
    right ^= (work << 8);

    right = ROLc(right, 1);
    work = (leftt ^ right) & 0xaaaaaaaaL;

    leftt ^= work;
    right ^= work;
    leftt = ROLc(leftt, 1);

    for (cur_round = 0; cur_round < 8; cur_round++) {
        work  = RORc(right, 4) ^ *keys++;
        leftt ^= SP7[work        & 0x3fL]
              ^ SP5[(work >>  8) & 0x3fL]
              ^ SP3[(work >> 16) & 0x3fL]
              ^ SP1[(work >> 24) & 0x3fL];
        work  = right ^ *keys++;
        leftt ^= SP8[ work        & 0x3fL]
              ^  SP6[(work >>  8) & 0x3fL]
              ^  SP4[(work >> 16) & 0x3fL]
              ^  SP2[(work >> 24) & 0x3fL];

        work = RORc(leftt, 4) ^ *keys++;
        right ^= SP7[ work        & 0x3fL]
              ^  SP5[(work >>  8) & 0x3fL]
              ^  SP3[(work >> 16) & 0x3fL]
              ^  SP1[(work >> 24) & 0x3fL];
        work  = leftt ^ *keys++;
        right ^= SP8[ work        & 0x3fL]
              ^  SP6[(work >>  8) & 0x3fL]
              ^  SP4[(work >> 16) & 0x3fL]
              ^  SP2[(work >> 24) & 0x3fL];
    }

    right = RORc(right, 1);
    work = (leftt ^ right) & 0xaaaaaaaaL;
    leftt ^= work;
    right ^= work;
    leftt = RORc(leftt, 1);
    work = ((leftt >> 8) ^ right) & 0x00ff00ffL;
    right ^= work;
    leftt ^= (work << 8);
    /* -- */
    work = ((leftt >> 2) ^ right) & 0x33333333L;
    right ^= work;
    leftt ^= (work << 2);
    work = ((right >> 16) ^ leftt) & 0x0000ffffL;
    leftt ^= work;
    right ^= (work << 16);
    work = ((right >> 4) ^ leftt) & 0x0f0f0f0fL;
    leftt ^= work;
    right ^= (work << 4);

    block[0] = right;
    block[1] = leftt;
}

 /**
    Initialize the LTC_DES block cipher
    @param key The symmetric key you wish to pass
    @param keylen The key length in bytes
    @param num_rounds The number of rounds desired (0 for default)
    @param skey The key in as scheduled by this function.
    @return CRYPT_OK if successful
 */
static int des_setup(const unsigned char *key, int keylen, int num_rounds, symmetric_key *skey)
{
    LTC_ARGCHK(key != NULL);
    LTC_ARGCHK(skey != NULL);

    if (num_rounds != 0 && num_rounds != 16) {
        return CRYPT_INVALID_ROUNDS;
    }

    if (keylen != 8) {
        return CRYPT_INVALID_KEYSIZE;
    }

    deskey(key, EN0, skey->des.ek);
    deskey(key, DE1, skey->des.dk);

    return CRYPT_OK;
}

/**
  Encrypts a block of text with LTC_DES
  @param pt The input plaintext (8 bytes)
  @param ct The output ciphertext (8 bytes)
  @param skey The key as scheduled
  @return CRYPT_OK if successful
*/
static int des_ecb_encrypt(const unsigned char *pt, unsigned char *ct, symmetric_key *skey)
{
    ulong32 work[2];
    LTC_ARGCHK(pt   != NULL);
    LTC_ARGCHK(ct   != NULL);
    LTC_ARGCHK(skey != NULL);
    LOAD32H(work[0], pt+0);
    LOAD32H(work[1], pt+4);
    desfunc(work, skey->des.ek);
    STORE32H(work[0],ct+0);
    STORE32H(work[1],ct+4);
    return CRYPT_OK;
}

/**
  Decrypts a block of text with LTC_DES
  @param ct The input ciphertext (8 bytes)
  @param pt The output plaintext (8 bytes)
  @param skey The key as scheduled
  @return CRYPT_OK if successful
*/
static int des_ecb_decrypt(const unsigned char *ct, unsigned char *pt, symmetric_key *skey)
{
    ulong32 work[2];
    LTC_ARGCHK(pt   != NULL);
    LTC_ARGCHK(ct   != NULL);
    LTC_ARGCHK(skey != NULL);
    LOAD32H(work[0], ct+0);
    LOAD32H(work[1], ct+4);
    desfunc(work, skey->des.dk);
    STORE32H(work[0],pt+0);
    STORE32H(work[1],pt+4);
    return CRYPT_OK;
}

/* $Source: /cvs/libtom/libtomcrypt/src/ciphers/des.c,v $ */
/* $Revision: 1.15 $ */
/* $Date: 2007/05/12 14:20:27 $ */

/**
  @file ccccrypt.c
  CommonCryptor compatibility
*/

#define kCCSuccess 0
#define kCCParamError -4300
#define kCCUnimplemented -4305
#define kCCEncrypt 0
#define kCCDecrypt 1
#define kCCAlgorithmDES 1
#define kCCOptionECBMode 2
#define kCCKeySizeDES 8

int cCCCrypt(int, int, int, const void *, size_t, const void *,
             const void *, size_t, void *, size_t, size_t *);

int cCCCrypt(int op, int alg, int opt, const void *key, size_t keylen,
             const void *iv, const void *in, size_t inlen,
	     void *out, size_t outsize, size_t *outlen)
{
  symmetric_key des;

  if ((op != kCCEncrypt && op != kCCDecrypt) || alg != kCCAlgorithmDES ||
      opt != kCCOptionECBMode)
    return kCCUnimplemented;
  if (!key || keylen != kCCKeySizeDES || iv || !in || inlen != 8 || !out ||
      outsize < 8 || !outlen)
    return kCCParamError;

  des_setup(key, 8, 0, &des);
  if (op == kCCEncrypt)
    des_ecb_encrypt(in, out, &des);
  else
    des_ecb_decrypt(in, out, &des);
  *outlen = 8;

  return kCCSuccess;
}
